core/src/main/java/io/parsingdata/metal/data/selection/ByOffset.java

```diff
@@ -19,53 +19,60 @@
 import static io.parsingdata.metal.Util.checkNotNull;
 
 import io.parsingdata.metal.data.ParseGraph;
-import io.parsingdata.metal.data.ParseGraphList;
 import io.parsingdata.metal.data.ParseItem;
+import io.parsingdata.metal.data.ParseItemList;
 import io.parsingdata.metal.data.ParseValue;
+import io.parsingdata.metal.token.Token;
 
 public final class ByOffset {
 
     private ByOffset() {}
 
-    public static boolean hasGraphAtRef(final ParseGraph graph, final long ref) {
-        return findRef(ByType.getGraphs(graph), ref) != null;
+    public static boolean hasRootAtRef(final ParseGraph graph, final Token definition, final long ref) {
+        return findRef(ByToken.getAllRoots(graph, definition), ref) != null;
     }
 
-    public static ParseGraph findRef(final ParseGraphList graphs, final long ref) {
-        checkNotNull(graphs, "graphs");
-        if (graphs.isEmpty()) { return null; }
-        final ParseGraph res = findRef(graphs.tail, ref);
-        if (res != null) { return res; }
-        if (graphs.head.containsValue() && graphs.head.getLowestOffsetValue().getOffset() == ref) {
-            return graphs.head;
+    public static ParseItem findRef(final ParseItemList items, final long ref) {
+        checkNotNull(items, "graphs");
+        if (items.isEmpty()) { return null; }
+        if (items.head.isValue() && items.head.asValue().getOffset() == ref) {
+            return items.head;
         }
-        return null;
+        if (items.head.isGraph() && containsLocalValue(items.head.asGraph()) && ByOffset.getLowestOffsetValue(items.head.asGraph()).getOffset() == ref) {
```

Comment:
Well, if we remove the check then we'll need a bunch of `null` checks around all places where the `getLowest*` methods recurse or call each other.

On the other hand, the whole logic has turned into six methods and is quite tangled. So I will have a look to see whether it can be simplified (with or without the check).